wrappers/services.go

```diff
@@ -42,46 +44,133 @@ import (
+}
+
+func newUserServiceClient(users []string, inter Interacter) (*userServiceClient, error) {
+	uids, err := usersToUids(users)
+	if err != nil {
+		return nil, err
+	}
+	var keys []int
+	for uid := range uids {
+		keys = append(keys, uid)
+	}
+	return &userServiceClient{
+		cli:   client.NewForUids(keys...),
+		inter: inter,
+	}, nil
+}
+
+func (c *userServiceClient) stopServices(services ...string) error {
 	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout.DefaultTimeout))
 	defer cancel()
-	failures, err := cli.ServicesStop(ctx, services)
+	failures, err := c.cli.ServicesStop(ctx, services)
 	for _, f := range failures {
-		inter.Notify(fmt.Sprintf("Could not stop service %q for uid %d: %s", f.Service, f.Uid, f.Error))
+		c.inter.Notify(fmt.Sprintf("Could not stop service %q for uid %d: %s", f.Service, f.Uid, f.Error))
 	}
 	return err
 }
 
-func startUserServices(cli *client.Client, inter Interacter, services ...string) error {
+func (c *userServiceClient) startServices(services ...string) error {
 	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout.DefaultTimeout))
 	defer cancel()
-	startFailures, stopFailures, err := cli.ServicesStart(ctx, services)
+	startFailures, stopFailures, err := c.cli.ServicesStart(ctx, services)
 	for _, f := range startFailures {
-		inter.Notify(fmt.Sprintf("Could not start service %q for uid %d: %s", f.Service, f.Uid, f.Error))
+		c.inter.Notify(fmt.Sprintf("Could not start service %q for uid %d: %s", f.Service, f.Uid, f.Error))
 	}
 	for _, f := range stopFailures {
-		inter.Notify(fmt.Sprintf("While trying to stop previously started service %q for uid %d: %s", f.Service, f.Uid, f.Error))
+		c.inter.Notify(fmt.Sprintf("While trying to stop previously started service %q for uid %d: %s", f.Service, f.Uid, f.Error))
+	}
+	return err
+}
+
+func (c *userServiceClient) restartServices(reload bool, services ...string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout.DefaultTimeout))
+	defer cancel()
+	var failures []client.ServiceFailure
+	var err error
+	if reload {
+		failures, err = c.cli.ServicesReloadOrRestart(ctx, services)
+	} else {
+		failures, err = c.cli.ServicesRestart(ctx, services)
+	}
+	for _, f := range failures {
+		c.inter.Notify(fmt.Sprintf("Could not restart service %q for uid %d: %s", f.Service, f.Uid, f.Error))
 	}
 	return err
 }
 
-func stopService(sysd systemd.Systemd, inter Interacter, scope snap.DaemonScope, svcs []string) error {
+func reloadOrRestartServices(sysd systemd.Systemd, cli *userServiceClient, reload bool, scope snap.DaemonScope, svcs []string) error {
+	switch scope {
+	case snap.SystemDaemon:
+		if reload {
+			if err := sysd.ReloadOrRestart(svcs); err != nil {
+				return err
+			}
+		} else {
+			if err := sysd.Restart(svcs); err != nil {
+				return err
+			}
+		}
+	case snap.UserDaemon:
+		if err := cli.restartServices(reload, svcs...); err != nil {
+			return err
```

Comment:
we don't seem to exercise the error paths here

Reply:
added a unit test to cover this path